pkg/operator/cluster/ceph/mon/mon.go

```diff
@@ -134,61 +107,61 @@ func (c *Cluster) Start() error {
 		return fmt.Errorf("failed to initialize ceph cluster info. %+v", err)
 	}
 
-	if len(c.clusterInfo.Monitors) < c.Size {
-		c.startMons()
-	} else {
-		// Check the health of a previously started cluster
-		if err := c.checkHealth(); err != nil {
-			logger.Warningf("failed to check mon health %+v. %+v", c.clusterInfo.Monitors, err)
-		}
-	}
-
-	return nil
+	return c.startMons()
 }
 
 func (c *Cluster) startMons() error {
-	// init the mons config
-	mons := c.initMonConfig(c.Size)
+	// create statefulset with replicas 0
+	c.startStatefulSet(0)
 
-	// Assign the pods to nodes
-	if err := c.assignMons(mons); err != nil {
-		return fmt.Errorf("failed to assign pods to mons. %+v", err)
-	}
+	mons := []*monConfig{}
 
-	// Start one monitor at a time
+	// Init the mons config
 	for i := len(c.clusterInfo.Monitors); i < c.Size; i++ {
-		// Init the mon IPs
-		if err := c.initMonIPs(mons[0 : i+1]); err != nil {
-			return fmt.Errorf("failed to init mon services. %+v", err)
-		}
+		monName := getMonNameForID(i)
 
-		// save the mon config after we have "initiated the IPs"
-		if err := c.saveMonConfig(); err != nil {
+		c.clusterInfo.MonMutex.Lock()
+		c.clusterInfo.Monitors[monName] = mon.ToCephMon(monName, c.getMonDNSEndpoint(), mon.DefaultPort)
+		// Save the mon config after we have initiated/set the mon config
+		if err := c.saveConfigChanges(); err != nil {
+			c.clusterInfo.MonMutex.Unlock()
 			return fmt.Errorf("failed to save mons. %+v", err)
 		}
+		c.clusterInfo.MonMutex.Unlock()
 
-		// make sure we have the connection info generated so connections can happen
-		if err := WriteConnectionConfig(c.context, c.clusterInfo); err != nil {
-			return err
-		}
+		mons = append(mons, &monConfig{Name: monName})
+		logger.Debugf("mon endpoints used are: %s", c.clusterInfo.MonEndpoints())
+		c.updateStatefulSet(int32(i + 1))
 
-		// Start pods
-		if err := c.startPods(mons[0 : i+1]); err != nil {
-			return fmt.Errorf("failed to start mon pods. %+v", err)
+		logger.Debugf("wait for mons to join (currently at id %d)", i)
+		c.waitForMonsToJoin(mons)
+
+		// get mon Pod IP and add it to endpoints
+		// TODO add a watch that updates the endpoints list on pod changes
+		podIP, err := c.getMonIP(monName)
+		if err != nil {
+			return fmt.Errorf("failed getting ip from mon pod %s. %+v", monName, err)
 		}
+		c.clusterInfo.MonMutex.Lock()
```

Comment:
perhaps we can implement this a bit without having to do several lockings. Having locks are essential for avoiding corruption in global data but they also have a performance factor.